custom_components/zwave_mqtt/__init__.py

```diff
@@ -63,60 +65,70 @@ async def mark_platform_loaded(platform):
 
     @callback
     def send_message(topic, payload):
-        _LOGGER.debug("sending message to topic %s", topic)
         mqtt.async_publish(hass, topic, json.dumps(payload))
 
     options = OZWOptions(send_message=send_message, topic_prefix=f"{TOPIC_OPENZWAVE}/")
     manager = OZWManager(options)
 
     for component in PLATFORMS:
-        hass.async_create_task(
-            hass.config_entries.async_forward_entry_setup(entry, component)
-        )
+        hass.async_create_task(hass.config_entries.async_forward_entry_setup(entry, component))
 
     @callback
     def async_node_added(node):
-        _LOGGER.debug("NODE ADDED: %s - node id %s", node, node.id)
+        _LOGGER.debug("[NODE ADDED] node_id: %s", node.id)
         data_nodes[node.id] = node
         data_values[node.id] = []
 
     @callback
     def async_node_changed(node):
-        _LOGGER.debug("NODE CHANGED: %s", node)
+        _LOGGER.debug("[NODE CHANGED] node_id: %s", node.id)
         data_nodes[node.id] = node
 
+    @callback
+    def async_node_removed(node):
+        _LOGGER.debug("[NODE REMOVED] node_id: %s", node.id)
+        # Remove entities belonging to this node
+        node_data_values = data_values[node.id]
+        for values in node_data_values:
+            if values.entity:
+                hass.async_create_task(values.entity.async_remove())
```

Comment:
We should signal the entity to remove itself with our dispatch helper. We should avoid using the entity api from outside the entity since that's fragile and it often conflate concerns.

The entity api is only valid after the entity has been added to home assistant. It's good to keep the device api and entity api separate to avoid entangling things.

Reply:
Thanks @MartinHjelmare can you give me a pointer about that ?

Reply:
Here's an example of how to use our dispatch helper from the tuya integration:
https://github.com/home-assistant/core/blob/01599d44f5ea8edb503b365336c03bf7b7240270/homeassistant/components/tuya/__init__.py#L97-L99

https://github.com/home-assistant/core/blob/01599d44f5ea8edb503b365336c03bf7b7240270/homeassistant/components/tuya/__init__.py#L122-L126

Reply:
But make the delete callback a coroutine instead and await `self.async_remove` inside.

Reply:
Thanks! I'll submit new PR's later today.